File: solutions/problem_2462_total_cost_to_hire_k_workers/solution.cpp
/*
 * Problem 2462: Total Cost to Hire K Workers
 * ========================================
 * Difficulty: Medium
 * Tags: Array, Two Pointers, Heap (Priority Queue), Simulation
 * Pattern: Two Heaps / Greedy
 *
 * Time Complexity:  O((k+c) log c)
 * Space Complexity: O(c)
 */

#include <queue>
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    int totalCost(vector<int>& costs, int k, int candidates) {
        // Heap/Priority Queue - O(n log k) time
        priority_queue<int, vector<int>, greater<int>> pq;
        for (int val : costs) {
            pq.push(val);
            if ((int)pq.size() > k)
                pq.pop();
        }
        return pq.empty() ? 0 : pq.top();
    }
};

File: solutions/problem_2462_total_cost_to_hire_k_workers/solution.py
"""
Problem 2462: Total Cost to Hire K Workers
========================================
Difficulty: Medium
Tags: Array, Two Pointers, Heap (Priority Queue), Simulation
Pattern: Two Heaps / Greedy

Time Complexity:  O((k+c) log c)
Space Complexity: O(c)
"""

import heapq

class Solution:
    def totalCost(self, costs, k, candidates):
        n = len(costs)
        left_heap = []
        right_heap = []
        l, r = 0, n - 1
        for i in range(candidates):
            if l <= r:
                heapq.heappush(left_heap, (costs[l], l))
                l += 1
        for i in range(candidates):
            if l <= r:
                heapq.heappush(right_heap, (costs[r], r))
                r -= 1
        total = 0
        for _ in range(k):
            left_val = left_heap[0] if left_heap else (float('inf'), -1)
            right_val = right_heap[0] if right_heap else (float('inf'), -1)
            if left_val <= right_val:
                cost, idx = heapq.heappop(left_heap)
                total += cost
                if l <= r:
                    heapq.heappush(left_heap, (costs[l], l))
                    l += 1
            else:
                cost, idx = heapq.heappop(right_heap)
                total += cost
                if l <= r:
                    heapq.heappush(right_heap, (costs[r], r))
                    r -= 1
        return total
